### PR TITLE
Add an option to create a logfile

### LINKED ISSUE #76: Log CLI runs to file to be able collect logs from users
_As a user who figured out an error in DX Scanner, I would like to have a log file where I can copy&paste the log result to a new issue on Github in DX Scanner repository._

**Definition of done:**
- all runs of DX Scanner are logged to a file in the root (or /tmp) folder
- the log contains all errors, debugs and basic project configuration
- the log has hidden secrets

Read [our Contribution guide](https://github.com/DXHeroes/dx-scanner/blob/master/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by vypxl**
## Description
Adds a cli option `-l --log` to let the program create a `dxscanner.log` file which contains all output and debugging information.

## Motivation and Context
Resolves #76

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] I have updated the documentation accordingly.
- [ ] I have added new practice to practice list in README.md.
- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Things to do
- [ ] Ensure that the logfile is flushed upon exit

### CODE REVIEW

**Comment by prokopsimek** [comment]
@vypxl Hi, I haven't found a possibility to flush the log at the process exit with file stream. There is a possibility to do it with `writeFileSync` but it could slow down the execution. (https://stackoverflow.com/questions/43643467/how-to-write-incrementally-to-a-text-file-and-flush-output)

What about to use any existing logger that probably has the problem resolved? (e.g. https://github.com/pinojs/pino)

**Comment by vypxl** [comment]
I think it's fixed now, I used the low level file descriptor api which can be used to close the file synchronously.

**Comment by prokopsimek** [comment]
@vypxl The logger works but I am sad with so many logger duplicities. Are you able to create a Logger class that does `console.log` either `logfile.log`?

It could be a class similar to `MyLogger` class for NestJS => https://docs.nestjs.com/techniques/logger#custom-implementation.

Thank you for your pull request!

**Comment by vypxl** [comment]
Do you mean that I should globally replace the `console.[log, error, ..] methods? I would rather convert all `console.*` calls to `debug.*` calls and intercept those. 

I also found out that I do not need the debug wrapper class I wrote, because `debug` allows the replacement of it's default logging handlers.

**Comment by prokopsimek** [comment]
@vypxl Ahh, okey, it sounds great. We may convert them to debug. ✅

**Comment by vypxl** [comment]
I did what I described before, is this good now?

**Review by Jakub-Vacek — APPROVED** [review]

File: src/reporters/HTMLReporter.ts [hunk 1]
```diff
@@ -35,7 +36,7 @@ export class HTMLReporter implements IReporter {
     else reportPath = <string>this.argumentsProvider.html;
 
     await this.fileSystemService.writeFile(path.resolve(process.cwd(), reportPath), reportHTML);
-    console.log('Report was saved to ' + reportPath);
+    debug.log('Report was saved to ' + reportPath);
```
Comment (prokopsimek) [inline src/reporters/HTMLReporter.ts:39]: `debugLog `?

File: src/reporters/JSONReporter.ts [hunk 2]
```diff
@@ -15,7 +16,7 @@ export class JSONReporter implements IReporter {
 
   async report(practicesAndComponents: PracticeWithContextForReporter[]): Promise<void> {
     const reportString = this.buildReport(practicesAndComponents);
-    console.log(inspect(reportString, { showHidden: false, depth: null }));
+    debug.log(inspect(reportString, { showHidden: false, depth: null }));
```
Comment (prokopsimek) [inline src/reporters/JSONReporter.ts:19]: `debugLog `?

File: src/reporters/FixReporter.ts [hunk 3]
```diff
@@ -26,8 +27,7 @@ export class FixReporter implements IReporter {
     practicesAndComponents: PracticeWithContextForReporter[],
     practicesAndComponentsAfterFix: PracticeWithContextForReporter[],
   ): Promise<void> {
-    const reportString = this.buildReport(practicesAndComponents, practicesAndComponentsAfterFix);
-    console.log(reportString);
+    debug.log(this.buildReport(practicesAndComponents, practicesAndComponentsAfterFix));
```
Comment (prokopsimek) [inline src/reporters/FixReporter.ts:30]: `debugLog `?

File: src/reporters/CLIReporter.ts [hunk 4]
```diff
@@ -25,8 +26,7 @@ export class CLIReporter implements IReporter {
   }
 
   async report(practicesAndComponents: PracticeWithContextForReporter[]): Promise<void> {
-    const reportString = this.buildReport(practicesAndComponents);
-    console.log(reportString);
+    debug.log(this.buildReport(practicesAndComponents));
```
Comment (prokopsimek) [inline src/reporters/CLIReporter.ts:29]: `debugLog `?

File: src/services/gitlab/GitLabService.ts [hunk 5]
```diff
@@ -24,13 +27,10 @@ import {
...
   Symlink,
   UserInfo,
-  Branch,
 } from '../git/model';
 import { VCSServicesUtils } from '../git/VCSServicesUtils';
 import { CustomAxiosResponse, GitLabClient, PaginationGitLabCustomResponse } from './gitlabClient/gitlabUtils';
-import { RepositoryConfig } from '../../scanner/RepositoryConfig';
-import _ from 'lodash';
-const debug = Debug('cli:services:git:gitlab-service');
+const d = debugLog('cli:services:git:gitlab-service');
```
Comment (prokopsimek) [inline src/services/gitlab/GitLabService.ts:33]: move to constructor?

File: src/lib/logfile.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  /**
+   * Writes the given value to the log file, followed by a newline.
+   * @param {*} args The values to write
+   * @memberof LogFile
+   */
+  public log(format: any, ...args: any[]) {
+    this.write(util.format(format, ...args) + '\n');
+  }
```
Comment (prokopsimek) [inline src/lib/logfile.ts:40]: We should redact private secrets somewhere here.

**Review by prokopsimek — CHANGES_REQUESTED** [review]
@adelkahomolova Review done. Please check that all debug commands uses `debugLog` function + there is a critical issue with redacting secrets.

File: src/reporters/FixReporter.ts (hunk 3 above)
Comment (adelkahomolova) [reply src/reporters/FixReporter.ts:30]: `debugLog()` doesn't log in terminal. `debug.log()` logs both in the terminal and in the file if the `-l` is used.

File: src/reporters/JSONReporter.ts (hunk 2 above)
Comment (adelkahomolova) [reply src/reporters/JSONReporter.ts:19]: debugLog() doesn't log in terminal. debug.log() logs both in the terminal and in the file if the -l is used.

File: src/reporters/HTMLReporter.ts (hunk 1 above)
Comment (adelkahomolova) [reply src/reporters/HTMLReporter.ts:39]: debugLog() doesn't log in terminal. debug.log() logs both in the terminal and in the file if the -l is used.

File: src/reporters/CLIReporter.ts (hunk 4 above)
Comment (adelkahomolova) [reply src/reporters/CLIReporter.ts:29]: debugLog() doesn't log in terminal. debug.log() logs both in the terminal and in the file if the -l is used.

**Pull request merged**